pkg/cmd/search/shared/searcher.go

```diff
@@ -0,0 +1,178 @@
+package shared
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"math"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/cli/cli/v2/pkg/search"
+)
+
+type searcher struct {
+	client *http.Client
+	host   string
+}
+
+func NewSearcher(client *http.Client, host string) *searcher {
+	return &searcher{
+		client: client,
+		host:   host,
+	}
+}
+
+func (s *searcher) Search(query search.Query) (search.Result, error) {
+	result := search.Result{}
+	path := fmt.Sprintf("https://api.%s/search/%s", s.host, query.Kind)
+	queryString := url.Values{}
+	q := strings.Builder{}
+	quotedKeywords := quoteKeywords(query.Keywords)
+	q.WriteString(strings.Join(quotedKeywords, " "))
+	for k, v := range query.Qualifiers.ListSet() {
+		v = quoteQualifier(v)
+		q.WriteString(fmt.Sprintf(" %s:%s", k, v))
+	}
+	queryString.Set("q", q.String())
+	if query.Order.IsSet() {
+		queryString.Set(query.Order.Key(), query.Order.String())
+	}
+	if query.Sort.IsSet() {
+		queryString.Set(query.Sort.Key(), query.Sort.String())
+	}
+	pages := math.Ceil((float64(query.Limit) / float64(100)))
+	for i := 1; i <= int(pages); i++ {
+		queryString.Set("page", strconv.Itoa(i))
+		remaining := query.Limit - (i * 100)
+		if remaining > 100 {
+			queryString.Set("per_page", "100")
+		} else if remaining <= 0 {
+			queryString.Set("per_page", strconv.Itoa(query.Limit))
+		} else {
+			queryString.Set("per_page", strconv.Itoa(remaining))
+		}
+		url := fmt.Sprintf("%s?%s", path, queryString.Encode())
+		req, err := http.NewRequest("GET", url, nil)
+		if err != nil {
+			return result, err
+		}
+		req.Header.Set("Content-Type", "application/json; charset=utf-8")
+		req.Header.Set("Accept", "application/vnd.github.v3+json")
+		resp, err := s.client.Do(req)
+		if err != nil {
+			return result, err
+		}
+		defer resp.Body.Close()
+		success := resp.StatusCode >= 200 && resp.StatusCode < 300
+		if !success {
+			return result, handleHTTPError(resp)
+		}
+		b, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return result, err
+		}
+		pageResult := search.Result{}
+		err = json.Unmarshal(b, &pageResult)
+		if err != nil {
+			return result, err
+		}
+		result.IncompleteResults = pageResult.IncompleteResults
+		result.TotalCount = pageResult.TotalCount
+		result.Items = append(result.Items, pageResult.Items...)
+	}
+	return result, nil
+}
+
+func (s *searcher) URL(query search.Query) string {
+	path := fmt.Sprintf("https://%s/search", s.host)
+	queryString := url.Values{}
+	queryString.Set("type", query.Kind)
+	if query.Order.IsSet() {
+		queryString.Set(query.Order.Key(), query.Order.String())
```

Comment:
Minor: I find it a smell that the same value `query.Order.Key()` is used as a key for both web-based search and API-based search. Just because they are the same today doesn't mean that this is guaranteed. I would strictly separate the logic for constructing a request for API search and the URL for web search, because they are entirely separate search APIs (only the formats of the `q` string are the same, because they are passed down to ElasticSearch).

Reply:
I hear your concern. My take is that while they are not guaranteed to be the same in the future, all the search endpoints are stable and being used in production today so it seems highly unlikely the `order` and `sort` parameters will get changed and if they do I trust there would be some sort of deprecation time period for us to change.